### PR TITLE
fix: adding new table POSTING_CATEGORIES and changing the rules to reflect requirements

### PR DESCRIPTION
**Pull request by SamRobinson75684**
…flect reuierments and updating tests

## Description

 adding new table POSTING_CATEGORIES and changing the rules to reflect requirements

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/rules/Breast_Screening_lookupRules.json [hunk 1]
```diff
@@ -31,14 +21,8 @@
...
       {
-        "RuleName": "58.CurrentPosting.NonFatal",
-        "LocalParams": [
-          {
-            "Name": "currentPosting",
-            "Expression": "newParticipant.CurrentPosting"
-          }
-        ],
-        "Expression": "string.IsNullOrEmpty(currentPosting) OR dbLookup.ValidateCurrentPosting(currentPosting)",
+        "RuleName": "36.CurrentPosting.NonFatal",
```
Comment (stephhou) [inline application/CohortManager/rules/Breast_Screening_lookupRules.json:24]: This is the one I messaged you about on slack. 58 has different conditions to 36, you can use 58 as part of 36 but they are looking up different current posting values.

File: tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs [hunk 2]
```diff
@@ -464,10 +465,10 @@ public async Task Run_ValidatePrimaryCareProvider_CreatesException(string record
 
     [TestMethod]
     [DataRow(Actions.New, "LDN", "20241101")]
-    [DataRow(Actions.Amended, "LDN", "20241101")]
-    [DataRow(Actions.Removed, "LDN", "20241101")]
-    [DataRow(Actions.Amended, null, "20241101")]
-    [DataRow(Actions.Amended, "LDN", null)]
+    [DataRow(Actions.New, "LDN", "20241101")]
+    [DataRow(Actions.New, "LDN", "20241101")]
```
Comment (MWClayson-NHS) [inline tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs:469]: There seems to be duplicate Data rows.

File: application/CohortManager/src/Functions/Shared/Data/Database/DbLookupValidationBreastScreening.cs [hunk 3]
```diff
@@ -106,28 +111,50 @@ public bool ValidateLanguageCode(string languageCode) {
...
                     return false;
                 }
+                return true;
             }
         }
     }
+
+    private bool validatePostingCategories(string postingCategory)
+    {
+        string[] allPossiblePostingCategories = ["ENGLAND", "IOM", "DMS"];
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/DbLookupValidationBreastScreening.cs:151]: Can you make this array a const at the top of the file?

File: tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs (hunk 2 above)
Comment (MWClayson-NHS) [reply tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs:469]: And will this throw an exception if its not a new record?

File: application/CohortManager/src/Functions/Shared/Data/Database/DbLookupValidationBreastScreening.cs [hunk 4]
```diff
@@ -106,28 +111,50 @@ public bool ValidateLanguageCode(string languageCode) {
...
                 {
                     while (reader.Read())
                     {
-                        return reader.GetInt32(0) == 1;
+                        postingCategory = reader["POSTING_CATEGORY"].ToString();
+                        isCurrentPostingInDB = reader.GetInt32(1) == 1;
                     }
+                }
+
+                if (currentPosting != null && !isCurrentPostingInDB && !validatePostingCategories(currentPosting)
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/DbLookupValidationBreastScreening.cs:139]: Can you split this up a bit there is a lot of information in this if which makes it hard to read.
Have a look at [this](https://github.com/zakirullin/cognitive-load?tab=readme-ov-file#complex-conditionals)

File: tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs (hunk 2 above)
Comment (SamRobinson75684) [reply tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs:469]: I tested this unit test this morning and it works with update and delete.

Comment (SamRobinson75684) [reply tests/ScreeningValidationServiceTests/LookupValidation/LookupValidationTests.cs:469]: we're just not longer considered with the record being an update

**Pull request closed without merging**